web_app/contract_tools/ekubo_research.py

```diff
@@ -0,0 +1,238 @@
+                )
+                response.raise_for_status()
+
+                quote_data = QuoteResponse.model_validate(response.json())
+                return self._find_best_route(quote_data, amount, min_profit_bps)
+
+        except (httpx.RequestError, ValueError) as e:
+            print(f"Error getting quote: {e}")
+            return None
+
+    def _find_best_route(self, quote_data: QuoteResponse, input_amount: int,
+                         min_profit_bps: int) -> Optional[Tuple[List[SwapRoute], int]]:
+        """
+        Select most profitable route from quote response
+        """
+        best_profit = 0
+        best_route = None
+        best_output = 0
+
+        for route_quote in quote_data.routes:
+            output_amount = int(route_quote.amount)
+            profit_bps = (output_amount - input_amount) * 10000 // input_amount
+
+            # Consider price impact (converted from percentage to bps)
+            price_impact_bps = int(Decimal(route_quote.priceImpact) * 100)
+            effective_profit_bps = profit_bps - abs(price_impact_bps)
+
+            if effective_profit_bps > best_profit and effective_profit_bps >= min_profit_bps:
+                best_profit = effective_profit_bps
+                best_route = [self._parse_route_node(node) for node in route_quote.route]
+                best_output = output_amount
+
+        if best_route:
+            return best_route, best_output
+        return None
+
+    def _parse_route_node(self, node: Dict) -> SwapRoute:
+        """Parse single route node from API response"""
+        return SwapRoute(
+            pool_key=PoolKey(
+                token0=node['pool_key']['token0'],
+                token1=node['pool_key']['token1'],
+                fee=int(node['pool_key']['fee'], 16),
+                tick_spacing=node['pool_key']['tick_spacing'],
+                extension=node['pool_key']['extension']
+            ),
+            sqrt_ratio_limit=int(node['sqrt_ratio_limit'], 16),
+            skip_ahead=int(node['skip_ahead'], 16)
+        )
+
+    def prepare_flash_loan_tx(self, route: List[SwapRoute],
+            borrow_amount: TokenAmount) -> Dict:
+        """
+        Prepare transaction for flash loan arbitrage
+            
+        Transaction flow:
+        1. Borrow tokens via flash accounting (no explicit borrow call needed)
+        2. Execute swaps through provided route
+        3. Verify profit
+        4. Repay loan automatically from swap proceeds
+        """
+        return {
+                'contract_address': self.core_contract,
+                'entry_point_selector': 'multi_route_swap',
+                'calldata': self._encode_multi_route_swap(route, borrow_amount)
+            }
+
+    def _encode_multi_route_swap(self, route: List[SwapRoute],
+            token_amount: TokenAmount) -> List[int]:
+        """Encode parameters for multihop_swap"""
+        calldata = [len(route)]  # Route length
```

Comment:
Can't find `multi_route_swap` in the contract's [ABI](https://starkscan.co/contract/0x00000005dd3d2f4429af886cd1a3b08289dbcea99a294197e9eb43b0e0325b4b#read-write-contract-sub-write), also all calls to the core require processing callback. Are you sure you want to use exactly this contract and not `Router` for example? Where did you find this information? Is there some source?